### ISSUE TITLE
System.Reflection.DispatchProxy inconsistently fails for internal interfaces

### ISSUE THREAD

**Issue by AArnott** [dotnet/runtime#30917]
An undocumented limitation in `DispatchProxy.Create` is that it *sometimes* throws for various configurations of visibility. This seems to be a broader issue than the more specific dotnet/runtime#25595.

These configurations work:

1. public interface, public base class.
1. internal interface, internal base class, within the same assembly. But only when this is the *first* proxy ever generated.

These configurations *don't* work:

1. public interface, internal base class.
1. internal interface, internal base class, if another assembly has already provided an internal interface to generate a proxy for.

There were other odd configurations that similarly failed.

Since I've done this kind of feature before, I have a suspicion that the generated `ProxyBuilder` assembly includes an [`System.Runtime.CompilerServices.IgnoresAccessChecksToAttribute`](https://github.com/microsoft/vs-mef/blob/21ca4f58145ae9e0e1560de287e112a608ed796f/src/Microsoft.VisualStudio.Composition/Reflection/SkipClrVisibilityChecks.cs#L159) for the assembly that defines the interface. 
The way I solved this is that each time I need to generate a proxy, [I evaluate all the assemblies that visibility checks must be skipped for](https://github.com/microsoft/vs-mef/blob/21ca4f58145ae9e0e1560de287e112a608ed796f/src/Microsoft.VisualStudio.Composition/Configuration/MetadataViewGenerator.cs#L114) and create a new dynamic assembly for each unique set that I get. So multiple proxies are supported, and we reuse dynamic assemblies where possible but when the set of dynamic assemblies exceed any previously discovered set we have to start a new assembly.

Adding more attributes to an existing dynamic assembly is ineffective because once you've loaded the first type from the assembly, the CLR caches these attributes and doesn't notice when new ones are added.

**Comment by steveharter** [comment]
Is this blocking certain scenarios, and are there workarounds? Trying to determine priority.

Do we need to emit into a specified assembly?

**Comment by AArnott** [comment]
It's blocking our use of this type.
Instead we resort to writing our own ref.emit code to dynamically generate proxies. Not a trivial task, but I've done it a few times before.
It may be possible to replace some of our custom proxies with this class once this issue is fixed.

**Comment by GrabYourPitchforks** [comment]
What configurations specifically are you looking to be supported? The public interface on an internal base class?

The reason for our questions is that we have a boatload of other reflection work planned for 5.0, and we're trying to figure out prioritization for everything. If this is moved above the cut line something else has to fall below the cut line. But maybe we can identify some small, self-contained nugget of work that wouldn't fix the problem for the general case, but which would make your scenario work. :)

**Comment by AArnott** [comment]
We would need both configurations that are broken as listed in the issue description to work in order to switch to this class.

FWIW I think you can prioritize this fairly low on your .NET 5 backlog. We've already completed our workaround and aren't likely to spend time switching to `DispatchProxy` even if you did fix it before .NET 5 anyway.

**Comment by steveharter** [comment]
Due to 5.0 schedule constraints, this is being moved to Future.

**Comment by krwq** [comment]
There are couple of issues here:
- we're missing the IgnoresAccessChecksToAttribute for the assembly of proxy base type (if it's not visible)
- adding new IgnoresAccessChecksToAttribute with Ref.Emit after first type is loaded doesn't get respected by the runtime

first one is an easy 1-line fix, for the second issue I can see 3 ways we could fix this:
- workaround by creating new assembly whenever we need to add new attribute (seems messy and also if users do similar code elsewhere they will also need to do that)
- fix in the runtime: fix this scenario - I'm not currently sure how to do that
- add new hidden attribute "ignore all checks for this assembly" - this one I have an idea how to do - note that adding this attribute after loading first type would have similar issue to the one above but most likely you'd want to do it once up front

I think the third solution I suggested has lots of benefits for code gen - first of all it makes code much simpler and there is no need to analyze which assemblies you need to add the attribute for. Second solution is something we should fix regardless but fix might be more involved (I didn't quite figure out how to fix this - seems like metadata in VM is out of date vs what it should be - I tried to quickly remove caching of attributes to see if it would fix the issue but it didn't)

**Comment by AArnott** [comment]
In verifying the fix, I used runtime version 6.0.0-preview.2.21105.12. 
From my original bug description that described two scenarios, I got mixed results:

> internal interface, internal base class, if another assembly has already provided an internal interface to generate a proxy for.

✅ Fixed

> public interface, internal base class.

❌ Still broken

**Repro steps**

Compile and run a C# console project with this source file:

```cs
using System;
using System.Reflection;

IFoo foo = DispatchProxy.Create<IFoo, MyProxy>();
Console.WriteLine(foo.Add(3, 5));

public interface IFoo
{
    int Add(int a, int b);
}

internal class MyProxy : DispatchProxy
{
    protected override object Invoke(MethodInfo targetMethod, object[] args)
    {
        if (targetMethod.Name == nameof(IFoo.Add))
        {
            return (int)args[0] + (int)args[1];
        }

        throw new NotImplementedException();
    }
}
```

**Expected output**

8

**Actual Output**

```
 C:\temp\DispatchProxyTest\dotnet-sdk-6.0.100-preview.1.21107.2-win-x64\dotnet.exe run -p .\DispatchProxyTest\
Unhandled exception. System.TypeLoadException: Access is denied: 'MyProxy'.
   at System.Reflection.Emit.TypeBuilder.CreateTypeNoLock()
   at System.Reflection.Emit.TypeBuilder.CreateType()
   at System.Reflection.DispatchProxyGenerator.ProxyBuilder.CreateType()
   at System.Reflection.DispatchProxyGenerator.GenerateProxyType(Type baseType, Type interfaceType)
   at System.Reflection.DispatchProxyGenerator.GetProxyType(Type baseType, Type interfaceType)
   at System.Reflection.DispatchProxyGenerator.CreateProxyInstance(Type baseType, Type interfaceType)
   at System.Reflection.DispatchProxy.Create[T,TProxy]()
   at <Program>$.<Main>$(String[] args) in C:\temp\DispatchProxyTest\DispatchProxyTest\Program.cs:line 4
```

**Comment by AArnott** [comment]
Ah, but all works great in 6.0.0-preview.3.21152.1. Re-closing.
